File: 2023/domaci/10-abecedni-poradi/volfmatous/main.js
const readline = require("readline");
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    terminal: false,
});

let linesRead = 0;
const lines = [];

async function readLine() {
    if (lines.length < 1) {
        for await (const line of rl) {
            lines.push(line);
        }
    }

    linesRead++;
    return lines[linesRead - 1];
}

main().then();

async function main() {
    const n = parseInt(await readLine());

    let descending = false;
    let descendingDecided = false;
    let previousWord = "";
    for (let i = 0; i < n; i++) {
        const currentWord = await readLine();

        const inOrderAscending = areWordsInOrder(previousWord, currentWord, false);
        const inOrderDescending = areWordsInOrder(previousWord, currentWord, true);

        previousWord = currentWord;

        if (descendingDecided) {
            if (descending ? !inOrderDescending : !inOrderAscending) {
                console.log("unordered");
                return;
            }
        } else {
            if (inOrderAscending && inOrderDescending) {
                continue;
            }

            descendingDecided = true;

            if (inOrderAscending) {
                descending = false;
                continue;
            }

            if (inOrderDescending) {
                descending = true;
                continue;
            }

            console.log("unordered");
            return;
        }
    }

    console.log("ordered");
}

function areWordsInOrder(word1, word2, descending) {
    let letterCaseInOrder = true;
    let letterCaseInOrderDecided = false;

    const shorterLength = Math.min(word1.length, word2.length);
    for (let i = 0; i < shorterLength; i++) {
        const char1 = word1[i];
        const char2 = word2[i];

        if (char1.toLowerCase() !== char2.toLowerCase()) {
            const alphabetAscendingOrder = char1.toLowerCase() < char2.toLowerCase();
            return descending ? !alphabetAscendingOrder : alphabetAscendingOrder;
        }

        if (letterCaseInOrderDecided) {
            continue;
        }

        if (char1.toLowerCase() === char1 && char2.toUpperCase() === char2) {
            letterCaseInOrder = true;
            letterCaseInOrderDecided = true;
            continue;
        }

        if (char1.toUpperCase() === char1 && char2.toLowerCase() === char2) {
            letterCaseInOrder = false;
            letterCaseInOrderDecided = true;
            continue;
        }
    }

    if (word1.length !== word2.length) {
        return word1.length < word2.length;
    }

    return letterCaseInOrder;
}

File: 2023/domaci/10-abecedni-poradi/volfmatous/main.py
import sys

def are_words_in_order(word1, word2, descending):
    letter_case_in_order = True
    letter_case_in_order_decided = False

    shorter_length = min(len(word1), len(word2))
    for i in range(shorter_length):
        char1 = word1[i]
        char2 = word2[i]

        if char1.lower() != char2.lower():
            alphabet_ascending_order = char1.lower() < char2.lower()
            return not alphabet_ascending_order if descending else alphabet_ascending_order

        if letter_case_in_order_decided:
            continue

        if char1.islower() and char2.isupper():
            letter_case_in_order = True
            letter_case_in_order_decided = True
            continue

        if char1.isupper() and char2.islower():
            letter_case_in_order = False
            letter_case_in_order_decided = True
            continue

    if len(word1) != len(word2):
        return len(word1) < len(word2)

    return letter_case_in_order


if __name__ == "__main__":
    n = int(input())

    descending = False
    descending_decided = False
    previous_word = ""
    for _ in range(n):
        current_word = input()

        in_order_ascending = are_words_in_order(previous_word, current_word, False)
        in_order_descending = are_words_in_order(previous_word, current_word, True)

        previous_word = current_word

        if descending_decided:
            if (descending and not in_order_descending) or (not descending and not in_order_ascending):
                print("unordered")
                sys.exit()
        else:
            if in_order_ascending and in_order_descending:
                continue

            descending_decided = True

            if in_order_ascending:
                descending = False
                continue

            if in_order_descending:
                descending = True
                continue

            print("unordered")
            sys.exit()

    print("ordered")
